rust/otel-arrow-rust/src/otap/batching.rs

```diff
@@ -0,0 +1,259 @@
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct Config {
+    /// Equivalent to `send_batch_size` in the go implementation.
+    size_trigger: NonZeroU64,
+
+    /// Equivalent to `timeout` in the go implementation.
+    time_trigger: Option<Duration>,
+
+    /// Equivalent to `send_batch_max_size` in the go implementation.
+    max_output_batch: Option<NonZeroU64>,
+}
+
+impl Config {
+    /// Construct a new configuration struct
+    #[must_use]
+    pub fn new(
+        size_trigger: NonZeroU64,
+        time_trigger: Option<Duration>,
+        max_output_batch: Option<NonZeroU64>,
+    ) -> Self {
+        Config {
+            size_trigger,
+            time_trigger,
+            max_output_batch,
+        }
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        let size_trigger = NonZeroU64::new(8192).expect("this is fine");
+        let time_trigger = Some(Duration::from_millis(200));
+        let max_output_batch = None;
+        Config {
+            size_trigger,
+            time_trigger,
+            max_output_batch,
+        }
+    }
+}
+
+/// I only exist to conveniently group together data needed by `do_the_thing`.
+struct BatcherState {
+    config: Config,
+    input_rx: UnboundedReceiver<OtapArrowRecords>, // but I'm keeping this one
+    output_tx: UnboundedSender<OtapArrowRecords>,  // and I'm keeping this one too
+    termination_rx: oneshot::Receiver<()>,
+    current: Vec<OtapArrowRecords>,
+}
+
+/// I'm the user facing component
+pub struct Batcher {
+    input_tx: UnboundedSender<OtapArrowRecords>, // I give this one away
+    output_rx: Option<UnboundedReceiver<OtapArrowRecords>>, // but I'm giving this one away
+    // We use an Option here because you can't clone receivers, so we need the ability to note that
+    // we don't have it anymore.
+    termination_tx: Option<oneshot::Sender<()>>,
+    _task_handle: JoinHandle<()>,
+}
+
+impl Batcher {
+    /// How to send `OtapArrowRecords` for batching.
+    #[must_use]
+    pub fn inbound(&self) -> UnboundedSender<OtapArrowRecords> {
+        self.input_tx.clone()
+    }
+
+    /// How to receive properly batched `OtapArrowRecords`
+    pub fn outbound(&mut self) -> Option<UnboundedReceiver<OtapArrowRecords>> {
+        self.output_rx.take()
+    }
+}
+
+impl Batcher {
+    /// Create a new `Batcher`!
+    pub async fn new(config: Config) -> Self {
+        let (input_tx, input_rx) = unbounded_channel();
+        let (output_tx, output_rx) = unbounded_channel();
+        let output_rx = Some(output_rx);
+        let (termination_tx, termination_rx) = oneshot::channel();
+        let termination_tx = Some(termination_tx);
+
+        let current = Vec::new();
+        let state = BatcherState {
+            config,
+            input_rx,
+            output_tx,
+            termination_rx,
+            current,
+        };
+        let _task_handle = tokio::task::spawn(update_state(state));
```

Comment:
For a first version it’s probably okay, but we shouldn’t have to create a task at this level. The BatchProcessor is already a task in itself that can, via the effect handler, request to be woken up on a recurring basis at regular intervals. I need to continue the review before I can go further with my feedback on this point.